### PR TITLE
Task4

### PR DESCRIPTION
**Pull request by Kylehinks97**
how is task 4 looking? :)

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -44,3 +41,43 @@ describe("TASK 3.5 --- GET - /api", () => {
       });
   });
 });
+describe.only("TASK 4 --- GET - /api/articles/:article_id", () => {
```
Comment (dappernerddesigns) [inline __tests__/app.test.js]: I know you've done a second commit removing this, be careful about the .only in your testing.

File: __tests__/app.test.js [hunk 2]
```diff
@@ -44,3 +41,43 @@ describe("TASK 3.5 --- GET - /api", () => {
...
+  it("GET - status: 200 - responds with JSON object containing correct article contents", () => {
+    return request(app)
+      .get("/api/articles/1")
+      .then((response) => {
+        console.log(response.body)
```
Comment (dappernerddesigns) [inline __tests__/app.test.js]: Make sure console logs are removed before prs

File: __tests__/app.test.js [hunk 3]
```diff
@@ -44,3 +41,43 @@ describe("TASK 3.5 --- GET - /api", () => {
...
+        expect(response.body.result.created_at).toBe("2020-07-09T19:11:00.000Z");
+        expect(response.body.result.votes).toBe(100);
+        expect(response.body.result.article_img_url).toBe('https://images.pexels.com/photos/158651/news-newsletter-newspaper-information-158651.jpeg?w=700&h=700');
+      });
+  });
+  it('GET - status: 400 - responds with error message explaining it was a bad request', () => {
+    return request(app)
+    .get(`/api/articles/nonsense`)
+    .then((response) => {
+      console.log(response);
```
Comment (dappernerddesigns) [inline __tests__/app.test.js]: see above, no console logs in production code.

File: __tests__/app.test.js [hunk 4]
```diff
@@ -44,3 +41,43 @@ describe("TASK 3.5 --- GET - /api", () => {
...
+      expect(response.statusCode).toBe(400)
+      expect(JSON.parse(response.text)).toEqual({ msg: "Bad Request" })
+    })
+  });
+});
+
+
+/*
+{
```
Comment (dappernerddesigns) [inline __tests__/app.test.js]: Make sure to remove comments.

File: __tests__/app.test.js [hunk 5]
```diff
@@ -44,3 +41,43 @@ describe("TASK 3.5 --- GET - /api", () => {
...
+        expect(response.body.result.article_img_url).toBe('https://images.pexels.com/photos/158651/news-newsletter-newspaper-information-158651.jpeg?w=700&h=700');
+      });
+  });
+  it('GET - status: 400 - responds with error message explaining it was a bad request', () => {
+    return request(app)
+    .get(`/api/articles/nonsense`)
+    .then((response) => {
+      console.log(response);
+      expect(response.statusCode).toBe(400)
+      expect(JSON.parse(response.text)).toEqual({ msg: "Bad Request" })
```
Comment (dappernerddesigns) [inline __tests__/app.test.js]: You shouldn't need to parse the error response. What you're sending back should be easily grabbed on the response body.

File: db/controllers.js [hunk 6]
```diff
@@ -12,15 +12,26 @@ exports.getTopics = (req, res) => {
...
 };
+exports.getArticleById = (req, res) => {
+  const id  = req.params.article_id
+  console.log(id);
+  selectArticleById(id)
+    .then((result) => {
+      res.status(200).send({result});
+  })
+    .catch((err) => {
+      if (err) console.log(err);
```
Comment (dappernerddesigns) [inline db/controllers.js]: There should not be any if statements or defined errors in this catch block. What you have written here is saying that no matter the error send back a 400. What if the article doesn't exist? That is not an invalid input. The error handling section of the notes has a very good example of what to do here.

File: db/models.js [hunk 7]
```diff
@@ -9,3 +9,12 @@ exports.selectTopics = () => {
       return result;
     });
 };
+exports.selectArticleById = (id) => {
+  return db.query(`
+  SELECT * FROM articles WHERE article_id = $1
+  `, [id])
+    .then((result) => {
+      const article = result.rows
+      return article[0]
```
Comment (dappernerddesigns) [inline db/models.js]: What would we tell the user if the article id is 999? Completely valid ID but not one in the database.

**Review by dappernerddesigns — CHANGES_REQUESTED** [review]
This ticket is missing a 404 test and the endpoints JSON has not been updated. Please address the comments and the missing test and resubmit this ticket.

File: db/app.js [hunk 8]
```diff
@@ -1,15 +1,31 @@
...
+    } else {
+      next(err);
+    }
+  });
 
 app.all("*", (req, res) => {
-    res.status(404).send("404 - Not Found")
-})
+  res.status(404).send("404 - Not Found");
+});
```
Comment (HannahBurks) [inline db/app.js:30]: We also want to add something to handle error 500 (server error) this would be the final error (you can't test for this)

**Review by HannahBurks — APPROVED** [review]
Lovely - feel free to merge away :D just look to add that 500 error and look at checking error codes on PSQL errors

**Review by HannahBurks — APPROVED** [review]
Lovely - feel free to merge away :D just look to add that 500 error and look at checking error codes on PSQL errors

File: db/app.js (hunk 8 above)
Comment (Kylehinks97) [reply db/app.js:30]: would that be something like another app.use if status(500) send a message about it, else next(err)?

**Pull request merged**